File: medication-tracker/backend/app/services/drug_interaction_service.py
import requests
from flask import current_app
import json
from datetime import datetime, timedelta
import logging

class DrugInteractionService:
    def __init__(self):
        self.base_url = "https://api.fda.gov/drug/label.json"
        self.interaction_cache = {}
        self.cache_duration = timedelta(days=7)  # Cache interaction data for 7 days

    def get_drug_interactions(self, drug_name):
        """
        Get drug interactions from the FDA API
        """
        try:
            # Check cache first
            cache_key = drug_name.lower()
            if cache_key in self.interaction_cache:
                cached_data = self.interaction_cache[cache_key]
                if datetime.now() - cached_data['timestamp'] < self.cache_duration:
                    return cached_data['data']

            # Query FDA API
            params = {
                'search': f'openfda.brand_name:"{drug_name}" OR openfda.generic_name:"{drug_name}"',
                'limit': 1
            }
            
            response = requests.get(self.base_url, params=params)
            response.raise_for_status()
            
            data = response.json()
            
            if 'results' not in data or not data['results']:
                return None
                
            result = data['results'][0]
            interactions = {
                'drug_interactions': result.get('drug_interactions', []),
                'warnings': result.get('warnings', []),
                'contraindications': result.get('contraindications', []),
                'precautions': result.get('precautions', [])
            }
            
            # Cache the result
            self.interaction_cache[cache_key] = {
                'data': interactions,
                'timestamp': datetime.now()
            }
            
            return interactions

        except requests.exceptions.RequestException as e:
            current_app.logger.error(f"Error fetching drug interactions: {str(e)}")
            return None

    def check_interaction(self, med1_name, med2_name):
        """
        Check for interactions between two medications
        """
        try:
            med1_data = self.get_drug_interactions(med1_name)
            med2_data = self.get_drug_interactions(med2_name)
            
            if not med1_data or not med2_data:
                return None
                
            interactions = []
            
            # Check if med2 is mentioned in med1's interactions
            for interaction in med1_data['drug_interactions']:
                if med2_name.lower() in interaction.lower():
                    interactions.append({
                        'severity': 'high',
                        'description': interaction
                    })
                    
            # Check if med1 is mentioned in med2's interactions
            for interaction in med2_data['drug_interactions']:
                if med1_name.lower() in interaction.lower():
                    interactions.append({
                        'severity': 'high',
                        'description': interaction
                    })
                    
            # Check warnings and precautions
            for warning in med1_data['warnings'] + med1_data['precautions']:
                if med2_name.lower() in warning.lower():
                    interactions.append({
                        'severity': 'medium',
                        'description': warning
                    })
                    
            for warning in med2_data['warnings'] + med2_data['precautions']:
                if med1_name.lower() in warning.lower():
                    interactions.append({
                        'severity': 'medium',
                        'description': warning
                    })
            
            return interactions if interactions else None

        except Exception as e:
            current_app.logger.error(f"Error checking drug interactions: {str(e)}")
            return None

    def get_detailed_info(self, drug_name):
        """
        Get detailed information about a drug
        """
        try:
            params = {
                'search': f'openfda.brand_name:"{drug_name}" OR openfda.generic_name:"{drug_name}"',
                'limit': 1
            }
            
            response = requests.get(self.base_url, params=params)
            response.raise_for_status()
            
            data = response.json()
            
            if 'results' not in data or not data['results']:
                return None
                
            result = data['results'][0]
            
            return {
                'brand_name': result.get('openfda', {}).get('brand_name', []),
                'generic_name': result.get('openfda', {}).get('generic_name', []),
                'description': result.get('description', []),
                'indications_and_usage': result.get('indications_and_usage', []),
                'dosage_and_administration': result.get('dosage_and_administration', []),
                'warnings': result.get('warnings', []),
                'drug_interactions': result.get('drug_interactions', []),
                'contraindications': result.get('contraindications', []),
                'adverse_reactions': result.get('adverse_reactions', []),
                'precautions': result.get('precautions', [])
            }

        except Exception as e:
            current_app.logger.error(f"Error fetching drug details: {str(e)}")
            return None

# Create singleton instance
drug_interaction_service = DrugInteractionService()

File: medication-tracker/backend/src/services/DrugInteractionService.ts
import { injectable, inject } from 'inversify';
import { Logger } from 'winston';
import axios, { AxiosError } from 'axios';
import NodeCache from 'node-cache';
import { TYPES } from '@/config/types.js';
import { ApiError } from '@/utils/errors.js';
import { auditLog } from '@/utils/audit.js';
import { monitorPerformance } from '@/utils/monitoring.js';
import { 
  DrugInteractionData,
  InteractionResult,
  InteractionSeverity,
  InteractionType,
  TimingInteraction,
  INTERACTION_CONSTANTS
} from '@/types/interactions.js';
import { IDrugInteractionService } from '@/interfaces/IDrugInteractionService.js';
import { differenceInHours } from 'date-fns';

@injectable()
export class DrugInteractionService implements IDrugInteractionService {
  private readonly baseUrl = 'https://api.fda.gov/drug/label.json';
  private readonly interactionCache: NodeCache;

  constructor(
    @inject(TYPES.Logger) private readonly logger: Logger
  ) {
    this.interactionCache = new NodeCache({
      stdTTL: INTERACTION_CONSTANTS.CACHE_DURATION_DAYS * 24 * 60 * 60, // Convert days to seconds
      maxKeys: INTERACTION_CONSTANTS.MAX_CACHE_SIZE
    });
  }

  @monitorPerformance('get_drug_interactions')
  public async getDrugInteractions(drugName: string): Promise<DrugInteractionData | null> {
    try {
      // Check cache first
      const cacheKey = `interactions_${drugName.toLowerCase()}`;
      const cachedData = this.interactionCache.get<DrugInteractionData>(cacheKey);
      if (cachedData) {
        await auditLog('drug_interaction', 'cache_hit', { drugName });
        return cachedData;
      }

      // Query FDA API
      const params = {
        search: `openfda.brand_name:"${drugName}" OR openfda.generic_name:"${drugName}"`,
        limit: 1
      };

      const response = await axios.get(this.baseUrl, {
        params,
        timeout: INTERACTION_CONSTANTS.API_TIMEOUT_MS
      });

      if (!response.data.results?.[0]) {
        await auditLog('drug_interaction', 'not_found', { drugName });
        return null;
      }

      const data = this.parseInteractionData(response.data.results[0]);
      this.interactionCache.set(cacheKey, data);
      await auditLog('drug_interaction', 'api_success', { drugName });
      return data;

    } catch (error) {
      this.logger.error('Error getting drug interactions:', error);
      if (error instanceof AxiosError && error.response?.status === 404) {
        return null;
      }
      throw error instanceof ApiError ? error : new ApiError('Failed to get drug interactions', 500);
    }
  }

  @monitorPerformance('check_interaction')
  public async checkInteraction(
    med1Name: string,
    med2Name: string
  ): Promise<InteractionResult[]> {
    try {
      const [med1Data, med2Data] = await Promise.all([
        this.getDrugInteractions(med1Name),
        this.getDrugInteractions(med2Name)
      ]);

      if (!med1Data || !med2Data) {
        throw new ApiError('One or both medications not found', 404);
      }

      const interactions: InteractionResult[] = [];

      // Check if med2 is mentioned in med1's interactions
      for (const interaction of med1Data.drugInteractions) {
        if (interaction.toLowerCase().includes(med2Name.toLowerCase())) {
          interactions.push(this.createInteractionResult(
            'severe',
            'drug_interaction',
            med1Name,
            med2Name,
            interaction
          ));
        }
      }

      // Check if med1 is mentioned in med2's interactions
      for (const interaction of med2Data.drugInteractions) {
        if (interaction.toLowerCase().includes(med1Name.toLowerCase())) {
          interactions.push(this.createInteractionResult(
            'severe',
            'drug_interaction',
            med2Name,
            med1Name,
            interaction
          ));
        }
      }

      // Check warnings and precautions from med1
      for (const warning of [...med1Data.warnings, ...med1Data.precautions]) {
        if (warning.toLowerCase().includes(med2Name.toLowerCase())) {
          interactions.push(this.createInteractionResult(
            'moderate',
            'warning',
            med1Name,
            med2Name,
            warning
          ));
        }
      }

      // Check warnings and precautions from med2
      for (const warning of [...med2Data.warnings, ...med2Data.precautions]) {
        if (warning.toLowerCase().includes(med1Name.toLowerCase())) {
          interactions.push(this.createInteractionResult(
            'moderate',
            'warning',
            med2Name,
            med1Name,
            warning
          ));
        }
      }

      await auditLog('drug_interaction', 'check_complete', {
        med1Name,
        med2Name,
        interactionCount: interactions.length
      });

      return interactions;
    } catch (error) {
      this.logger.error('Error checking drug interaction:', error);
      throw error instanceof ApiError ? error : new ApiError('Failed to check drug interaction', 500);
    }
  }

  @monitorPerformance('check_timing_interaction')
  public async checkTimingInteraction(
    med1Name: string,
    med1Time: Date,
    med2Name: string,
    med2Time: Date
  ): Promise<TimingInteraction | null> {
    try {
      const hoursDifference = Math.abs(differenceInHours(med1Time, med2Time));
      
      if (hoursDifference < INTERACTION_CONSTANTS.MIN_HOURS_BETWEEN_DOSES) {
        const [med1Data, med2Data] = await Promise.all([
          this.getDrugInteractions(med1Name),
          this.getDrugInteractions(med2Name)
        ]);

        if (!med1Data || !med2Data) {
          return null;
        }

        return {
          med1Name,
          med2Name,
          timeDifference: hoursDifference,
          recommendation: `Consider spacing out ${med1Name} and ${med2Name} by at least ${INTERACTION_CONSTANTS.MIN_HOURS_BETWEEN_DOSES} hours`,
          severity: hoursDifference < 2 ? 'high' : 'moderate'
        };
      }

      return null;
    } catch (error) {
      this.logger.error('Error checking timing interaction:', error);
      throw error instanceof ApiError ? error : new ApiError('Failed to check timing interaction', 500);
    }
  }

  private createInteractionResult(
    severity: InteractionSeverity,
    type: InteractionType,
    sourceMed: string,
    targetMed: string,
    description: string
  ): InteractionResult {
    return {
      severity,
      type,
      sourceMed,
      targetMed,
      description,
      timestamp: new Date()
    };
  }

  private parseInteractionData(rawData: any): DrugInteractionData {
    return {
      drugInteractions: rawData.drug_interactions || [],
      warnings: rawData.boxed_warning || [],
      precautions: rawData.precautions || [],
      contraindications: rawData.contraindications || []
    };
  }
}
